### PR TITLE
feat: wrap ShareProof in a ResultShareProof

### PR DESCRIPTION
**Pull request by rach-id**
## Description

Contributes to https://github.com/celestiaorg/celestia-core/issues/1306

### CODE REVIEW

**Comment by rach-id** [comment]
CI will be fixed with: https://github.com/celestiaorg/celestia-core/pull/1347

File: light/rpc/client.go
Comment (rootulp) [inline light/rpc/client.go:584]:
```suggestion
// Deprecated: use ProveSharesV2 instead.
```

Ref: 
- https://go.dev/wiki/Deprecated
- https://stackoverflow.com/questions/7849663/how-do-you-mark-code-as-deprecated-in-go

File: light/rpc/client.go
Comment (rootulp) [inline light/rpc/client.go:597]: nit: I find this go doc difficult to read.

> calls rpcclient#ProveSharesV2 method

that seems like implementation details that the caller doesn't need to know about.

> and returns an NMT proof for a set of shares, defined by `startShare` and `endShare`, to the corresponding rows.

It's not clear to me if one proof is returned or multiple proofs. I _think_ that this would return one NMT proof per row that is occupied by the share range [startShare, endShare). Additionally, I think the NMT proof proves that some shares exist in an NMT with a particular row root.

> // Then, a binary merkle inclusion proof from the latter rows to the data root.

The "latter rows" here refer to the rows that are occupied by the share range?

IMO this Go doc could be made clearer by stating something like:

```suggestion
// ProveSharesV2 returns a result share proof that includes multiple proofs to prove that a range of shares exist in a particular block. See ____ for docs on how to verify the share proof.
```

and then the `___` docs are located on a Go function like: `ResultShareProof.Proof.Verify()`.

File: rpc/client/http/http.go
Comment (rootulp) [inline rpc/client/http/http.go:554]:
```suggestion
// Deprecated: use ProveSharesV2 instead.
```

File: rpc/client/interface.go
Comment (rootulp) [inline rpc/client/interface.go:87]:
```suggestion
	// Deprecated: use ProveSharesV2 instead.
```

File: rpc/core/types/responses.go
Comment (rootulp) [inline rpc/core/types/responses.go:278]: [question] why not `ShareProof`?

```suggestion
	ShareProof types.ShareProof `json:"share_proof"`
```

**Review by rootulp — CHANGES_REQUESTED** [review]
Please use `Deprecated: `

Ref: https://go.dev/wiki/Deprecated

File: rpc/core/types/responses.go
Comment (rach-id) [reply rpc/core/types/responses.go:278]: no reason, if you prefer `share_proof` I can switch

File: light/rpc/client.go
Comment (rach-id) [reply light/rpc/client.go:597]:
```suggestion
// ProveSharesV2 returns a share to data root inclusion proof.
```
Let's just simplify it like this.

File: rpc/core/types/responses.go
Comment (rootulp) [reply rpc/core/types/responses.go:278]: Since this `ShareProof` is actually composed of a bunch of other proofs, I feel like `ShareProof` helps differentiate it from the several other types of proof inside. I don't have a strong preference so defer to you.

Comment (rach-id) [reply rpc/core/types/responses.go]: [7252ab8](https://github.com/celestiaorg/celestia-core/pull/1351/commits/7252ab8d2af2ae652d814bb0776cda1f499f70ed)

File: light/rpc/client.go
Comment (rootulp) [reply light/rpc/client.go:597]: IMO that isn't clear because this proof is for a range of shares not an individual share.

Comment (rach-id) [reply light/rpc/client.go:597]:
```suggestion
// ProveSharesV2 returns a shares to data root inclusion proof.
```
What do you think?

**Review by evan-forbes — DISMISSED** [review]

File: light/rpc/client.go
Comment (rootulp) [reply light/rpc/client.go:597]: idk why I still find that phrasing confusing. My alternative suggestion would be:

```
// ProveSharesV2 returns a proof of inclusion for a share range to the data root. Note: this proof is composed of multiple proofs.
```

File: rpc/core/tx.go
Comment (rootulp) [inline rpc/core/tx.go:217]: This Go doc seems incomplete because ProveShares also contains a binary Merkle tree proof for the relevant row roots to the data root.

File: rpc/core/types/responses.go
Comment (rootulp) [inline rpc/core/types/responses.go:276]: This Go doc doesn't actually describe what an "API proof response of a set of shares"

```suggestion
// ResultShareProof is an API response that contains a ShareProof.
```

**Review by rootulp — DISMISSED** [review]
IMO the Go docs here should be improved but I'm not going to block on it.

File: rpc/core/tx.go
Comment (rach-id) [reply rpc/core/tx.go]: [30a7bd4](https://github.com/celestiaorg/celestia-core/pull/1351/commits/30a7bd4cf995cdea86d0cbdc304682c097d4a11d)

**Review by rootulp — APPROVED** [review]

File: rpc/openapi/openapi.yaml
Comment (staheri14) [inline rpc/openapi/openapi.yaml:1005]: [Suggestion] As a reviewer, verifying the OpenAPI specifications can be challenging and error-prone. I recommend using tools like [go-swagger](https://github.com/go-swagger/go-swagger) to generate the specifications automatically hence simplifying the process for developers and reviewers, providing greater confidence in the generated code.

File: light/rpc/client.go
Comment (staheri14) [inline light/rpc/client.go:601]: Mind adding some comments about the values permitted for the start and end shares?

File: light/rpc/client.go
Comment (staheri14) [inline light/rpc/client.go:595]: [nit]
```suggestion
// ProveSharesV2 returns a proof of inclusion for a share range to the data root of the given height.
```

File: rpc/client/local/local.go
Comment (staheri14) [inline rpc/client/local/local.go:226]: [question] Why don't we initially define the type of `height` as int64?

**Review by staheri14 — APPROVED** [review]
LGTM!

**Comment by rach-id** [comment]
@staheri14 Will address the feedback in an upcoming PR. Thanks for the review

File: rpc/openapi/openapi.yaml
Comment (rach-id) [reply rpc/openapi/openapi.yaml:1005]: I am not sure whether the openapi specs in core was generated using go-swagger. We should investigate which tool they used.

For verifying that the specs are valid, I usually use https://editor.swagger.io/ and test the methods myself using an RPC endpoint.

File: rpc/client/local/local.go
Comment (rach-id) [reply rpc/client/local/local.go:226]: it's a choice that has been done when we started building Blobstream two years ago, and we never got around to change it to `int64`.

**Pull request merged**
